crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -578,6 +592,87 @@ where
         })
     }
 
+    /// Returns true iff it is a primitive, an ID, a String, or an option/vector of a valid type
+    pub fn is_entry_primitive_type(&self, param_ty: &Type) -> Result<bool, ExecutionError> {
+        let mut stack = vec![param_ty];
+        while let Some(cur) = stack.pop() {
+            match cur {
+                Type::Signer => return Ok(false),
+                Type::Reference(_) | Type::MutableReference(_) | Type::TyParam(_) => {
+                    return Ok(false)
+                }
+                Type::Bool
+                | Type::U8
+                | Type::U16
+                | Type::U32
+                | Type::U64
+                | Type::U128
+                | Type::U256
+                | Type::Address => (),
+                Type::Vector(inner) => stack.push(&**inner),
+                Type::Struct(idx) => {
+                    let Some(s) = self.session.get_struct_type(*idx) else {
+                    invariant_violation!("Loaded struct not found")
+                };
+                    let resolved_struct = get_struct_ident(&s);
+                    if ![RESOLVED_SUI_ID, RESOLVED_ASCII_STR, RESOLVED_UTF8_STR]
+                        .contains(&resolved_struct)
+                    {
+                        return Ok(false);
+                    }
+                }
+                Type::StructInstantiation(idx, targs) => {
+                    let Some(s) = self.session.get_struct_type(*idx) else {
+                    invariant_violation!("Loaded struct not found")
+                };
+                    let resolved_struct = get_struct_ident(&s);
+                    // is option of a primitive
+                    let is_valid = resolved_struct == RESOLVED_STD_OPTION && targs.len() == 1;
+                    if !is_valid {
+                        return Ok(false);
+                    }
+                    stack.extend(targs)
+                }
+            }
+        }
+        Ok(true)
+    }
+
+    pub fn join(&self, t1: ValueType, t2: ValueType) -> Result<Option<ValueType>, ExecutionError> {
+        Ok(match (t1, t2) {
+            // everything joins with itValueType
+            (t1, t2) if t1 == t2 => Some(t1),
+            // AnyPrimitive joins with primitive types types. It is otherwise unsafe to join with
+            // AnyPrimitive as it is just a series of BCS bytes
+            // TODO dev-inspect, allow AnyPrimitive to join with any type?
+            (ValueType::AnyPrimitive, ValueType::Loaded { ty, abilities })
+            | (ValueType::Loaded { ty, abilities }, ValueType::AnyPrimitive) => {
```

Comment:
This case scares me a bit but maybe feels okay? Basically its a question of if we want to let `CallArg::Pure` appear in `MakeMoveVec`